### PR TITLE
Add more integration tests

### PR DESCRIPTION
**Pull request by marco-ballario**
This PR adds DeepSpeech and grcov as git submodules, their json metrics files and 2 integration tests.
The tests compare rca output for the submodules with their json files previously generated.
To do that I used the APIs to process files in a directory concurrently introduced here: https://github.com/mozilla/rust-code-analysis/pull/709
It partly resolves issue #389

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[submodule "tests/data/cpp/DeepSpeech"]
+	path = tests/data/cpp/DeepSpeech
+	url = https://github.com/mozilla/DeepSpeech.git
+[submodule "tests/data/rust/grcov"]
+	path = tests/data/rust/grcov
+	url = https://github.com/mozilla/grcov.git
```
Comment (marco-c) [inline .gitmodules:6]: I think it'd be preferable to avoid submodules, and simply clone the repos at the time we have to run the tests. This way we don't force everyone to clone them even if they don't need to run the intergration tests, and the repository stays smaller.

**Comment by marco-c** [comment]
Might be better to avoid storing all the expected JSON files in the repo, but run RCA before/after a change like we do for mozilla-central integration tests.
Another option would be to store all the JSON in a separate repository that we clone at the time of the integration tests, this way we don't need to store all these JSON files in this repository.

File: .gitmodules (hunk 1 above)
Comment (marco-ballario) [reply .gitmodules:6]: I will try to replace submodules with git clone commands at the beginning of my tests. In that way repositories are downloaded as first thing while running the tests (if not already present locally). Together with the repositories, I will also try to clone a specific version of RCA during my tests, so I can use it to produce the json files.

Thanks a lot for the suggestions!

Comment (marco-c) [reply .gitmodules:6]: > I will also try to clone a specific version of RCA during my tests, so I can use it to produce the json files.

We can also store them in a separate repo, it might actually be better:
1) We avoid running RCA twice (before and after)
2) We have a history of how the values change over time with the evolution of RCA

Comment (marco-ballario) [reply .gitmodules:6]: You're right, I can try that!
Where should I create the repository containing all the json files?

Comment (marco-c) [reply .gitmodules:6]: You can create it in your personal GitHub account for now

Comment (marco-ballario) [reply .gitmodules:6]: Michele gave me the permission to create the repository at https://github.com/SoftengPoliTo, so I will try to create it there.
Thanks for the support!

**Review by Luni-4 — APPROVED** [review]
Seems fine to me! Thank you!

File: src/tools.rs [hunk 2]
```diff
@@ -232,7 +232,7 @@ pub fn guess_language<P: AsRef<Path>>(buf: &[u8], path: P) -> (Option<LANG>, Str
 pub(crate) fn remove_blank_lines(data: &mut Vec<u8>) {
     let count_trailing = data.iter().rev().take_while(|&c| *c == b'\n').count();
     if count_trailing > 0 {
-        data.truncate(data.len() - count_trailing + 1);
```
Comment (marco-c) [inline src/tools.rs:235]: Why is this change required?

File: src/concurrent_files.rs [hunk 3]
```diff
@@ -92,6 +92,7 @@ where
         }
         if path.is_dir() {
             for entry in WalkDir::new(path)
+                .sort_by_file_name()
```
Comment (marco-c) [inline src/concurrent_files.rs:95]: Why is this change required?

File: tests/test.rs [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+    compare_rca_output_with_files(
+        "v0.9.3",
+        "https://github.com/mozilla/DeepSpeech.git",
+        "./tests/repositories/DeepSpeech",
+        &["*.cc", "*.cpp", "*.h", "*.hh"],
+    );
+}
+
+#[test]
+fn test_grcov() {
```
Comment (marco-c) [inline tests/test.rs:253]: We could use the Rust standard library ("library" subfolder in https://github.com/rust-lang/rust, but you can just use the whole repo since you have the `.rs` glob) for testing Rust, this way we make sure we're good with latest Rust language features (they are definitely going to use them in the Rust standard library).

File: src/tools.rs (hunk 2 above)
Comment (marco-ballario) [reply src/tools.rs:235]: During my tests I noticed that few DeepSpeech source files generated different json files in Windows and Linux. Thanks to @Luni-4 and @lu-zero help we determined that the cause of those differences were the end of line formats at the end of few downloaded source files. This line should fix this issue and allow the generation of the same json files both in Windows and in Linux.

File: src/concurrent_files.rs (hunk 3 above)
Comment (marco-ballario) [reply src/concurrent_files.rs:95]: We found that file access order in a directory depends on the operating system policy. This line should define a unique deterministic exploration order valid for every operating system.

File: tests/test.rs (hunk 4 above)
Comment (marco-ballario) [reply tests/test.rs:253]: Thank you for the review! Do you think I should substitute grcov tests with Rust standard library tests? Or I can leave grcov tests and simply add Rust standard library tests?

File: src/concurrent_files.rs (hunk 3 above)
Comment (marco-c) [reply src/concurrent_files.rs:95]: Why do we need a deterministic exploration order?

File: tests/test.rs (hunk 4 above)
Comment (marco-c) [reply tests/test.rs:253]: Both options work for me, you choose ;)

File: src/tools.rs (hunk 2 above)
Comment (marco-c) [reply src/tools.rs:235]: Unless I'm mistaken, after your change this line is basically doing nothing (truncating `data` to its full length) and the `remove_blank_lines` function wouldn't actually remove blank lines.
@calixteman why was `remove_blank_lines` necessary?

Comment (Luni-4) [reply src/tools.rs:235]: I added that function to remove the blank lines at the end of a file, such that we don't consider them for the `Blank` metric counting. But on Windows new lines are identified by `/r/n`, so we have a different `data` length when we compare the source files parsed on Linux with the ones parsed on Windows. One solution could be leaving blank lines at the end of a file, as implemented in this PR (I agree with this approach because blank lines at the end of a file should be considered), otherwise we can find `\r\n` on Windows

File: src/concurrent_files.rs (hunk 3 above)
Comment (marco-ballario) [reply src/concurrent_files.rs:95]: You are right. It seems this line is not strictly necessary for the tests to work. Source files can be read in any order since the json file path is deducted directly from the source file path. Comparisons for a single file will never be affected by this change. I will remove this change. Thank you for pointing this out!

File: src/tools.rs (hunk 2 above)
Comment (marco-c) [reply src/tools.rs:235]: > One solution could be leaving blank lines at the end of a file, as implemented in this PR

So we could just remove the `remove_blank_lines` function entirely?

File: src/concurrent_files.rs (hunk 3 above)
Comment (marco-c) [reply src/concurrent_files.rs:95]: No problem! I actually didn't know that it wasn't necessary, it's good I asked :)

File: src/tools.rs (hunk 2 above)
Comment (Luni-4) [reply src/tools.rs:235]: Looking more carefully the code, `remove_blank_lines` removes blank lines when `count_trailing > 0` leaving a single new line at the end of a file. So my first solution is not feasible, the second one is correct instead

File: src/tools.rs [hunk 5]
```diff
@@ -229,13 +229,17 @@ pub fn guess_language<P: AsRef<Path>>(buf: &[u8], path: P) -> (Option<LANG>, Str
     }
 }
 
+/// Replaces \n and \r ending characters with a single generic \n
 pub(crate) fn remove_blank_lines(data: &mut Vec<u8>) {
-    let count_trailing = data.iter().rev().take_while(|&c| *c == b'\n').count();
+    let count_trailing = data
+        .iter()
+        .rev()
+        .take_while(|&c| (*c == b'\n' || *c == b'\r'))
```
Comment (marco-c) [inline src/tools.rs:237]: This is going to count Windows newlines twice.
On Linux you would have `Ciao\nCiao\n\n`, and so count would be 2.
On Windows you would have `Ciao\r\nCiao\r\n\r\n`, and so count would be 4.

Comment (marco-c) [reply src/tools.rs:237]: Forget it, the code after is removing the characters and not the lines.

**Review by marco-c — APPROVED** [review]
